doc/developer/design/20230418_subscribe_output.md

```diff
@@ -0,0 +1,282 @@
+table](#update-table) for more examples. If PROGRESS is set we also
+return `mz_progressed` as usual after `mz_state` and each progress
+message nulls out all the key and value columns.
+
+Using `ENVELOPE UPSERT` when there is more than one live value per key
+can be confusing. The upserts generated have no way to express which
+value was deleted for a given key.
+
+`ENVELOPE DEBEZIUM`: Similarly to `ENVELOPE UPSERT`, takes a list of
+columns to interpret as a key and within each distinct timestamp emits
+debezium upserts with a `before` and `after` state. The output columns
+are reordered as `(mz_timestamp, mz_state, k1, k2, .., before_v1,
+before_v2, .., after_v1, after_v2, ..)`. There is no `mz_diff`
+column. When there is an addition, `mz_state` is set to `"insert"` and
+the `before` values are all set to `NULL`. When there is an update,
+`mz_state` is set to `"upsert"`, `before` has the previous contents,
+and `after` has the current values. When there is a deletion,
+`mz_state` is set to `"delete"`, the `before` values are set to the
+previous values and the `after` values are all `NULL`.
+
+Similar to `ENVELOPE UPSERT` the `ENVELOPE DEBEZIUM` is confusing when
+there is more than one live value per key because new values may
+either appear as inserts or upserts. See the [update
+table](#update-table) for some example of confusing cases like rows 4
+and 5.
+
+If PROGRESS is set, the key, `before_v*` and `after_v*` are all set to
+`NULL` an addition `mz_progressed` column is inserted after
+`mz_state`.
+
+Today only one output modifier can be used at a time. `ENVELOPE
+DEBEZIUM` and `ENVELOPE UPSERT` are conflicting directions. `WITHIN
+TIMESTAMP ORDER BY` and `ENVELOPE DEBEZIUM` are hard to use together
+since `DEBEZIUM` produces two copies of the value and it would be
+difficult to distinguish between them. In theory, `WITHIN TIMESTAMP
+ORDER BY` and `ENVELOPE UPSERT` could be used together if the client
+wanted a specific sort order out of subscribe. There are few cases
+where that's useful so I propose disallowing it until there is client
+demand.
+
+### Update table:
+[update-table]: #update-table
+
+| Updates | `ENVELOPE UPSERT` | `ENVELOPE DEBEZIUM` |
+| --- | --- | --- |
+| `(k, v1, +1)` | `(k, v1)` | `(k, NULL, v1)` |
+| `(k, v1, +1)`, `(k, v2, +1)` | `(k, v1)`, `(k, v2)` | `(k, NULL, v1)`, `(k, NULL, v2)` |
+| `(k, v1, -1)`, `(k, v2, +1)` | `(k, v2)` | `(k, v1, v2)` |
+| `(k, v1, -1)`, `(k, v2, +1)`, `(k, v3, +1)` | `(k, v2)`, `(k, v3)` | either `(k, v1, v2)`, `(k, NULL, v3)` or `(k, v1, v3)`, `(k, NULL, v2)` |
+| `(k, v1, -1)`, `(k, v2, -1)`, `(k, v3, +1)` | `(k, v3)` | either `(k, v1, NULL)`, `(k, v2, v3)` or `(k, v2, NULL)`, `(k, v1, v3)` |
+| `(k, v1, -1)`, `(k, v2, -1)`, `(k, v3, +1)`, `(k, v4, +1)` | `(k, v3)`, `(k, v4)` | `(k, v1, v3)`, `(k, v2, v4)` or `(k, v2, v3)`, `(k, v1, v4)` |
+| `(k, v1, -1)` | `(k, NULL)` | `(k, v1, NULL)` |
+| `(k, v1, -1)`, `(k, v2, -1)` | `(k, NULL)` | `(k, v1, NULL)`, `(k, v2, NULL)` |
```

Comment:
Where is `mz_state` in these example?

Reply:
added. the value of mz_state corresponds to where the NULL is. mz_state is neccesary to disambiguate (1) a "normal" value of all nulls and (2) no value columns